mcp/mcp-cli/src/main/java/software/amazon/smithy/java/mcp/cli/ProcessIoProxy.java

```diff
@@ -0,0 +1,277 @@
+     */
+    private static Thread createForwardingThread(
+            Process process,
+            InputStream input,
+            OutputStream output,
+            String name,
+            String errorMessage,
+            boolean closeOutput,
+            AtomicBoolean running
+    ) {
+        return Thread.ofVirtual()
+                .name(name)
+                .start(() -> {
+                    try {
+                        byte[] buffer = new byte[BUFFER_SIZE];
+                        int bytesRead;
+                        while (running.get() && process.isAlive() && (bytesRead = input.read(buffer)) != -1) {
+                            output.write(buffer, 0, bytesRead);
+                            output.flush();
+                        }
+                    } catch (IOException e) {
+                        if (running.get()) {
+                            LOG.error(errorMessage, e);
+                        }
+                    } finally {
+                        if (closeOutput) {
+                            try {
+                                output.close();
+                            } catch (IOException e) {
+                                LOG.error("Error closing stream", e);
+                            }
+                        }
+                    }
+                });
+    }
+
+    /**
+     * Starts the proxy, launching the subprocess and beginning to forward stdin/stdout.
+     *
+     * @throws RuntimeException If there is an error starting the process
+     */
+    public synchronized void start() {
+        if (!running.get()) {
+            return;
+        }
+        try {
+            process = processBuilder.start();
+
+            // Thread to forward input to process
+            inputThread = createForwardingThread(
+                    process,
+                    inputStream,
+                    process.getOutputStream(),
+                    "process-stdin-forwarder",
+                    "Error forwarding input to process",
+                    true, // Close the process input stream when done
+                    running);
+
+            // Thread to forward process stdout to output
+            outputThread = createForwardingThread(
+                    process,
+                    process.getInputStream(),
+                    outputStream,
+                    "process-stdout-forwarder",
+                    "Error forwarding process stdout",
+                    false, // Don't close the output stream
+                    running);
+
+            // Thread to forward process stderr to error
+            errorThread = createForwardingThread(
+                    process,
+                    process.getErrorStream(),
+                    errorStream,
+                    "process-stderr-forwarder",
+                    "Error forwarding process stderr",
+                    false, // Don't close the error stream
+                    running);
+
+            running.set(true);
```

Comment:
The threads you spawn above here may begin running before you set this to true, causing them to instantly exit when they observe `running` as false. You should use compareAndSet on line 194.